src/server/pkg/storage/fileset/writer.go

```diff
@@ -141,76 +103,57 @@ func (w *Writer) nextIdx(idx *index.Index) error {
 
 func (w *Writer) checkPath(p string) error {
 	if w.idx != nil {
-		if w.idx.Path > p {
-			return errors.Errorf("can't write path (%s) after (%s)", p, w.idx.Path)
+		if w.idx.Path == p {
+			return errors.Errorf("cannot write same path (%s) twice", p)
 		}
-		parent := parentOf(p)
-		if w.idx.Path < parent {
-			return errors.Errorf("cannot write path (%s) without first writing parent (%s)", p, parent)
+		if w.idx.Path > p {
+			return errors.Errorf("cannot write path (%s) after (%s)", p, w.idx.Path)
 		}
 	}
 	return nil
 }
 
 // Delete creates a delete operation for a file.
-func (w *Writer) Delete(p string) error {
+// TODO: Check path order.
```

Comment:
All operations should operate on paths in order.  So we would have a single `lastPath` field for the whole Writer?

Reply:
That is not currently true. Additive operations are ordered with respect to other additive operations, but not with respect to deletive operations (and vice versa). The unordered writer has its own notion of an additive and deletive set and it is serialized additive first followed by deletive. I suppose we could merge them, but that seems unnecessary seeing as they are written to separate index writers. I think we might end drawing a clearer distinction between additive / deletive operations rather than having them operate on the same writer.

Reply:
I think we should add checks for this.  The Writer isn't threadsafe or anything so it should just be a straightforward conditional check. A lot depends on the ordering here being correct.